File: packages/types/transaction.ts
export interface TransactionItem {
  id: string;
  workspaceId: string;
  transactionId: string;
  name: string;
  brand?: string | null;
  quantity?: string | null;
  unit?: string | null;
  unitPrice?: string | null;
  amount: string;
  categoryId?: string | null;
  notes?: string | null;
  createdAt: string;
  updatedAt: string;
  deletedAt?: string | null;
  category?: { id: string; name: string } | null;
}

export interface Transaction {
  id: string;
  workspaceId: string;
  walletId: string;
  toWalletId?: string | null;
  categoryId?: string | null;
  amount: string; // Decimal is string in JS
  date: string;
  type: string; // 'income' | 'expense' | 'transfer' | 'transfer-in' | 'transfer-out'
  name?: string | null;
  description?: string | null;
  createdAt: string;
  updatedAt: string;
  isReady: boolean;
  isExported: boolean;
  assignedUserId?: string | null;
  deletedAt?: string | null;
  attachmentIds?: string[];
  attachments?: {
    id: string;
    name: string;
    size: number;
    type: string;
  }[];
  wallet?: { id: string; name: string };
  toWallet?: { id: string; name: string } | null;
  category?: { id: string; name: string } | null;
  user?: {
    id: string;
    name: string | null;
    email: string;
    profile_picture: string | null;
  } | null;
  items?: TransactionItem[];
}

File: apps/native/lib/domain/models/transaction.dart
import 'package:equatable/equatable.dart';
import 'package:flutter/foundation.dart';
import 'package:oewang/domain/models/money.dart';

/// Transaction kind as encoded by the API
/// (`income | expense | transfer | transfer-in | transfer-out`).
enum TransactionType {
  income,
  expense,
  transfer,
  transferIn,
  transferOut;

  static TransactionType fromWire(String raw) {
    switch (raw) {
      case 'income':
        return TransactionType.income;
      case 'expense':
        return TransactionType.expense;
      case 'transfer':
        return TransactionType.transfer;
      case 'transfer-in':
        return TransactionType.transferIn;
      case 'transfer-out':
        return TransactionType.transferOut;
      default:
        return TransactionType.expense;
    }
  }

  String get wire => switch (this) {
    TransactionType.income => 'income',
    TransactionType.expense => 'expense',
    TransactionType.transfer => 'transfer',
    TransactionType.transferIn => 'transfer-in',
    TransactionType.transferOut => 'transfer-out',
  };
}

/// Reference shape used inside [Transaction] for related entities.
@immutable
class NamedRef extends Equatable {
  const NamedRef({required this.id, required this.name});
  final String id;
  final String name;
  @override
  List<Object?> get props => [id, name];
}

@immutable
class Transaction extends Equatable {
  const Transaction({
    required this.id,
    required this.type,
    required this.amount,
    required this.date,
    required this.walletId,
    this.toWalletId,
    this.categoryId,
    this.name,
    this.description,
    this.wallet,
    this.toWallet,
    this.category,
  });

  final String id;
  final TransactionType type;
  final Money amount;

  /// Calendar date the user assigned to the transaction (YYYY-MM-DD on the
  /// wire). Time-of-day is intentionally dropped.
  final DateTime date;

  final String walletId;
  final String? toWalletId;
  final String? categoryId;

  final String? name;
  final String? description;

  final NamedRef? wallet;
  final NamedRef? toWallet;
  final NamedRef? category;

  bool get isIncome => type == TransactionType.income;
  bool get isExpense => type == TransactionType.expense;

  @override
  List<Object?> get props => [
    id,
    type,
    amount,
    date,
    walletId,
    toWalletId,
    categoryId,
    name,
    description,
    wallet,
    toWallet,
    category,
  ];
}
